### PR TITLE
feature: Adding support for Multi Worker Mirrored Strategy in TF estimator

### PR DESCRIPTION
**Pull request by Lokiiiiii**
*Issue #, if available:*
Resolves https://github.com/aws/sagemaker-tensorflow-training-toolkit/issues/391

*Description of changes:*
SageMaker Tensorflow Training ToolKit now supports Multi Worker Mirrored Strategy through cluster initialization.
These changes allow the cluster init to be invoked through the TF Estimator's distribution parameter.

*Testing done:*
PR validation tests

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I certify that the changes I am introducing will be backword compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [x] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [x] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/sagemaker/tensorflow/estimator.py [hunk 1]
```diff
@@ -137,6 +137,19 @@ def __init__(
                         To find a complete list of parameters for SageMaker model parallelism,
                         see :ref:`sm-sdk-modelparallel-general`.
 
+                **To enable Multi Worker Mirrored Strategy:**
```
Comment (mufaddal-rohawala) [inline src/sagemaker/tensorflow/estimator.py:140]: would this change be needed for other frameworks?

File: src/sagemaker/tensorflow/estimator.py [hunk 2]
```diff
@@ -137,6 +137,19 @@ def __init__(
...
+
+                    .. code:: python
+
+                        {
+                            "multi_worker_mirrored_strategy": {
+                                "enabled": True
+                            }
+                        }
+
+                    To learn more, see `TensorFlow Distributed Training
```
Comment (mchoi8739) [inline src/sagemaker/tensorflow/estimator.py:150]:
```suggestion
                    This distribution strategy option is available for TensorFlow 2.9 and later in
                    the SageMaker Python SDK v2.xx.yy and later.
                    To learn more about the mirrored strategy for TensorFlow, 
                    see `TensorFlow Distributed Training
```

Comment (mchoi8739) [reply src/sagemaker/tensorflow/estimator.py:150]: please update the SM PySDK version when the release cut is being finalized.

File: src/sagemaker/tensorflow/estimator.py [hunk 3]
```diff
@@ -137,6 +137,19 @@ def __init__(
...
+                    <https://www.tensorflow.org/guide/distributed_training>`_.
```
Comment (mchoi8739) [inline src/sagemaker/tensorflow/estimator.py:151]:
```suggestion
                    <https://www.tensorflow.org/guide/distributed_training>`_ in the *TensorFlow documentation*.
```

**Review by mchoi8739 — CHANGES_REQUESTED** [review]

File: src/sagemaker/tensorflow/estimator.py (hunk 1 above)
Comment (Lokiiiiii) [reply src/sagemaker/tensorflow/estimator.py:140]: No, this distribution mechanism is specific to TensorFlow.

File: src/sagemaker/estimator.py [hunk 4]
```diff
@@ -3429,6 +3449,12 @@ def _distribution_configuration(self, distribution):
                     "dataparallel"
                 ].get("custom_mpi_options", "")
 
+        if "multi_worker_mirrored_strategy" in distribution:
+            mwms_enabled = distribution.get("multi_worker_mirrored_strategy").get("enabled", False)
+            if mwms_enabled:
+                self._validate_mwms_config()
```
Comment (pinaraws) [inline src/sagemaker/estimator.py:3455]: Please also add checks to make sure this option is not enabled together with other incompatible options.

Comment (Lokiiiiii) [reply src/sagemaker/estimator.py]: Done

**Review by pinaraws — APPROVED** [review]

**Review by knikure — COMMENTED** [review]
/bot run pr

File: src/sagemaker/estimator.py [hunk 5]
```diff
@@ -3066,6 +3064,34 @@ def _validate_and_set_debugger_configs(self):
                     )
                     self.debugger_hook_config = False
 
+    def _validate_mwms_config(self, distribution):
```
Comment (mufaddal-rohawala) [inline src/sagemaker/estimator.py:3067]: Estimator.py file is becoming too lengthy. Considering the importance of this module and the way it's growing is there room to move this validation logic outside of this module?

**Comment by mufaddal-rohawala** [comment]
Hey @Lokiiiiii the related integ/unit tests are failing can you fix them?

**Comment by mufaddal-rohawala** [comment]
@Lokiiiiii is this PR still active?

**Review by trajanikant — COMMENTED** [review]
/bot run pr, slow-tests

**Review by trajanikant — COMMENTED** [review]
/bot run pr, slow-tests

**Review by mufaddal-rohawala — APPROVED** [review]

**Review by mufaddal-rohawala — COMMENTED** [review]
/bot run slow-tests

File: tests/conftest.py [hunk 6]
```diff
@@ -607,6 +602,22 @@ def inf_instance_family(inf_instance_type):
     return "_".join(inf_instance_type.split(".")[0:2])
 
 
+@pytest.fixture(scope="session")
+def imagenet_train_set(request, sagemaker_session, tmpdir_factory):
+    """
+    Copies the Imagenet dataset from the bucket it's hosted in to the local bucket in the test region.
+    Due to licensing issues, access to this dataset is controlled through an allowlist
```
[the remark this answers is not in the document]
Comment (Lokiiiiii) [reply tests/conftest.py]: Removed unused fixture. This is no longer a concern

**Review by mufaddal-rohawala — APPROVED** [review]

**Review by mufaddal-rohawala — COMMENTED** [review]
/bot run slow-tests, notebook-tests

**Pull request merged**